crates/subspace-networking/src/create.rs

```diff
@@ -56,6 +58,84 @@ const KADEMLIA_RECORD_REPLICATION_FACTOR: NonZeroUsize =
 // Defines a replication factor for Kademlia on get_record operation.
 // "Good citizen" supports the network health.
 const KADEMLIA_CACHING_FACTOR_ON_GET_RECORDS: u16 = 3;
+const YAMUX_MAX_STREAMS: usize = 256;
+
+/// Base limit for number of concurrent tasks initiated towards Kademlia.
+///
+/// Kademlia has 32 substream as a hardcoded constant, we leave 2 for auxiliary internal functions
+/// like periodic random walk.
+///
+/// We restrict this so we don't exceed number of incoming streams for single peer, but this value
+/// will be boosted depending on number of connected peers.
+const KADEMLIA_BASE_CONCURRENT_TASKS: usize = 30;
+/// Above base limit will be boosted by specified number for every peer connected starting with
+/// second peer, such that it scaled with network connectivity, but the exact coefficient might need
+/// to be tweaked in the future.
+const KADEMLIA_CONCURRENT_TASKS_BOOST_PER_PEER: usize = 1;
+/// Base limit for number of any concurrent tasks.
+///
+/// We configure total number of streams per connection to 256. Here we assume half of them might be
+/// incoming and half outgoing, we also leave a small buffer of streams just in case.
+///
+/// We restrict this so we don't exceed number of streams for single peer, but this value will be
+/// boosted depending on number of connected peers.
+const GLOBAL_BASE_CONCURRENT_TASKS: usize = 120;
+/// Above base limit will be boosted by specified number for every peer connected starting with
+/// second peer, such that it scaled with network connectivity, but the exact coefficient might need
+/// to be tweaked in the future.
+const GLOBAL_CONCURRENT_TASKS_BOOST_PER_PEER: usize = 2;
+const SEMAPHORE_MAINTENANCE_INTERVAL: Duration = Duration::from_secs(5);
+
+async fn maintain_semaphore_permits_capacity(
```

Comment:
Guess semaphores are not best fit for this scenario, where the max limit can grow/shrink as peers come and go. Hence we need this complexity to hold excess permits, etc

How do you feel about this:
1. Maintain `current_max_limit, current_usage`.  These two are protected by async Mutex/CV.
2. `current_max_limit` can be updated with the boost_per_peer from `handle_swarm_event()` during connection open/close. 
3. Senders: wait for permits to be freed, add to `current_usage`
4. From `drop()` of the permit: bump up `current_usage`, signal any waiters. But, any excess permits if the max limit shrinked are not returned to the free pool. This would need spawning async task from the dtor unfortunately, as rust doesn't support await from drop yet.

Happy to make a branch off your PR and make a draft, if that helps

Reply:
It is a bit hacky, I agree, but not too much (for my taste).

I found CVs counter-intuitive to read in most cases and I feel like we'll implement essentially the same mechanism (practically speaking) in the end. It might be an interesting exercise, but I'm not sure it'll actually be a lot of value in that.

Supporting what we need here (resizing, dropping of unfinished acquisition) is probably quire a bit of effort.

Feel free to try unless you have anything more valuable to do.

Reply:
OK, looks like tokio doesn't have CondVar, so that won't work :-( But found sem has a `forget()` API: https://docs.rs/tokio/1.23.0/tokio/sync/struct.SemaphorePermit.html#method.forget

This can help avoid the reserved_permits stash, which would simplify this quite a bit:

```
if total_permits < expected_total_permits {
     semaphore.add_permits(expected_total_permits - total_permits);
} else {
     let num_excess = total_permits - expected_total_permits;
     // acquire()/forget() num_excess permits
}
```
Not a big deal: it may take a while for the changes to take effect (in the shrink case), as the sem acquisition in this path doesn't get higher priority (unless done from the dtor as in the other approach)

Reply:
We can use `forget`, but that'll mean we'll permanently increase internal capacity of the semaphore every time we call `forget`. And while we'll probably never reach https://docs.rs/tokio/1.23.0/tokio/sync/struct.Semaphore.html#associatedconstant.MAX_PERMITS on 64-bit platform, we can on 32-bit platform. It'll take a while, but when you hit it, it'll be hard to debug.

Reply:
yeah you are right.

I still feel this path can be simplified (took a bit to figure out this shrinking business), let me try out the idea I mentioned above off your branch.